pennylane/capture/custom_primitives.py

```diff
@@ -30,10 +33,140 @@ class PrimitiveType(Enum):
     TRANSFORM = "transform"
 
 
+def _make_hashable(obj: Any) -> Any:
+    """Convert potentially unhashable objects to hashable equivalents for JAX 0.7.0+.
+
+    JAX 0.7.0 requires all primitive parameters to be hashable. This helper converts
+    common unhashable types (list, dict, slice) to hashable tuples.
+
+    Args:
+        obj: Object to potentially convert to hashable form
+
+    Returns:
+        Hashable version of the object
+    """
+    if isinstance(obj, slice):
+        return (obj.start, obj.stop, obj.step)
+
+    # First, check if the object is already hashable
+    try:
+        hash(obj)
+        return obj
+    except TypeError:
+        pass
+
+    # Import here to avoid circular dependency and only when needed
+    # pylint: disable=import-outside-toplevel,too-many-return-statements
+    import jax
+    import numpy as np
+
+    # Check if obj is a JAX tracer - these are already hashable, don't convert
+    # Must check before Array check since tracers can also be Array instances
+    if isinstance(obj, jax.core.Tracer):
+        return obj
+
+    # Convert arrays (JAX and NumPy) to nested tuples for hashability
+    # Must check before list check since tolist() returns lists
+    if hasattr(jax, "Array") and isinstance(obj, jax.Array):
+        # Recursively convert nested lists from tolist() to tuples
+        return _make_hashable(obj.tolist())
+    if isinstance(obj, np.ndarray):
+        # Recursively convert nested lists from tolist() to tuples
+        return _make_hashable(obj.tolist())
+    if isinstance(obj, list):
+        return tuple(_make_hashable(item) for item in obj)
+    if isinstance(obj, dict):
+        # Sort by str(key) to handle non-comparable keys (e.g., class types like ABCCaptureMeta)
+        # This ensures consistent ordering without requiring keys to implement __lt__
+        return tuple(
+            sorted(((k, _make_hashable(v)) for k, v in obj.items()), key=lambda x: str(x[0]))
+        )
+    return obj
+
+
+def _restore_slice(obj: Any) -> Any:
+    """Restore slice objects from hashable tuple representation.
+
+    This is the inverse of _make_hashable for slice objects. When QmlPrimitive.bind()
+    converts slices to tuples for hashability, primitive implementations need to
+    convert them back.
+
+    Args:
+        obj: Object that may be a tuple representation of a slice
+
+    Returns:
+        slice object if obj is a 3-tuple with (start, stop, step), otherwise returns obj unchanged
+    """
+    if isinstance(obj, tuple) and len(obj) == 3:
+        # Tuples representing slices have exactly 3 elements (start, stop, step)
+        # Check if it's likely a slice by checking if elements look like slice components
+        # (None or integers for start/stop/step)
+        if all(x is None or isinstance(x, int) for x in obj):
+            return slice(*obj)
+    return obj
```

Comment:
I'm really skeptical that we actually need a helper function for this.  We know what values are supposed to be slices and which ones are not.  We can just do `slice(*arg)`. No need to abstract 7 characters.